test/test_autograd.py

```diff
@@ -12926,6 +12926,409 @@ def run_test(fn):
+
+            # Expected result: a.grad = (grad_out + 1) * 2 = 4
+            self.assertEqual(a.grad, torch.ones_like(a) * 4)
+
+        for _ in range(2):
+            test()
+
+    def _test_consumer_to_single_producer_case_3_correctness(
+        self, non_default_ambient_stream
+    ):
+        #                          Device    Stream
+        # Consumer (MulBackward):  cuda:0    s0
+        # Producer              :  cuda:1    cuda:1 default
+        # Gradient              :  cuda:0    cuda:0 default
+        class Producer(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, x):
+                # The node's canonical stream is the current stream
+                # of the device of the first output.
+                ctx.node_stream = torch.cuda.current_stream(1)
+                return x.to("cuda:1")
+
+            @staticmethod
+            def backward(ctx, gO):
+                out = gO.to("cuda:0")
+                with torch.cuda.device(0):
+                    torch.cuda._sleep(NUM_GPU_CYCLES_IN_ONE_SEC // 2)
+                # It's the node's responsibility to sync back to its canonical stream.
+                out.add_(1)
+                ctx.node_stream.wait_stream(torch.cuda.current_stream(0))
+                return out
+
+        def test():
+            self.synchronize_all_devices()
+            self.assert_all_streams_default()
+
+            with torch.cuda.device(0):
+                s0 = torch.cuda.Stream()
+                s1 = torch.cuda.Stream()
+            default_stream_0, default_stream_1 = self.get_default_streams()
+
+            # Ensure consumer node happens on non-default stream so that
+            # when FuncBackward produces a gradient on a default stream
+            # a sync is necessary.
+            with torch.cuda.stream(s0):
+                a = torch.ones(256, 256, requires_grad=True, device="cuda")
+                b = a * 2
+
+            default_stream_0.wait_stream(s0)
+            out = Producer.apply(b)
+
+            with torch.autograd.grad_mode.set_multithreading_enabled(False):
+                s1.wait_stream(default_stream_0)
+                ctx = (
+                    torch.cuda.stream(s1)
+                    if non_default_ambient_stream
+                    else contextlib.nullcontext()
+                )
+                with ctx:
+                    out.sum().backward()
+
+            self.synchronize_all_devices()
+
+            # Expected result: a.grad = (grad_out + 1) * 2 = 4
+            self.assertEqual(a.grad, torch.ones_like(a) * 4)
+
+        for _ in range(2):
+            test()
+
+    # This fails because we currently sync to the default stream
+    @unittest.skipIf(not TEST_CUDA, "requires CUDA")
+    @unittest.skipIf(torch.cuda.device_count() < 2, "GPU count is less than 2")
+    @unittest.expectedFailure
+    def test_consumer_to_single_producer_case_3_correctness_non_default_ambient_stream(
```

Comment:
> I'd expect test like this to fail today https://github.com/pytorch/pytorch/pull/142097#issuecomment-2567161794, did you add it?

@ngimel Yeah, made sure to include a case where the backward is run under non-default stream here. It fails as expected today.